Add company OR history

jcoveney opened the pull request:
Building on the work in https://github.com/tobymao/18xx/pull/210

Closes https://github.com/tobymao/18xx/issues/215

This extends the spreadsheet to have an OR history for each company

In time, we will convert toby to the usefulness of this tab :)

<img width="717" alt="Screen Shot 2020-05-01 at 23 30 12" src="https://user-images.githubusercontent.com/830087/80817407-f287d480-8c03-11ea-9749-5195b618cf65.png">

jcoveney commented:
Per discussion with toby, will add information for the last set of ORs to the cards

tobymao commented on lib/engine/corporation.rb:
```diff
@@ -15,7 +15,7 @@ class Corporation
 
     attr_accessor :ipoed, :par_price, :share_price, :tokens
-    attr_reader :companies, :coordinates, :min_price, :name, :full_name, :logo, :trains, :color
+    attr_reader :companies, :coordinates, :min_price, :name, :full_name, :logo, :trains, :color, :or_revenue_history
```
rename to revenue_history

jcoveney replied:
will do. renaming things -- the bane of dynamically typed languages :)

tobymao commented on lib/engine/corporation.rb:
```diff
@@ -99,6 +100,11 @@ def corporation?
       true
     end
 
+    def push_revenue_history!(revenue, round_num)
```
how about just add_revenue

jcoveney replied:
2 hard problems in computer science. will do

tobymao commented on lib/engine/corporation.rb:
```diff
@@ -99,6 +100,11 @@ def corporation?
     end
 
+    def push_revenue_history!(revenue, round_num)
+      @or_revenue_history.push([]) if round_num == 1
```
use << intsead of push

jcoveney replied:
ah, ruby :)

tobymao commented on assets/js/view/spreadsheet.rb:
```diff
@@ -31,7 +31,24 @@ def render_table
       # TODO: consider adding train availability
     end
 
+    def render_or_history_titles(arbitrary_corporation)
```
render_history(corporation)

jcoveney replied:
but that is not what this is. this generates the OR1.1, OR2.1, OR2.2 etc information for the title

it could be render_history_titles though, or something else

tobymao replied:
sounds good

tobymao commented on assets/js/view/spreadsheet.rb:
```diff
@@ -40,6 +57,7 @@ def render_title
           h(:th, { attrs: { colspan: 2 } }, 'Prices'),
           h(:th, { attrs: { colspan: 4 } }, 'Corporation'),
           h(:th, { style: { width: '20px' } }, ''),
+          h(:th, { attrs: { colspan: or_history_titles.length } }, 'OR History'),
```
size

jcoveney replied:
ruby *shakes fist* def will do

tobymao commented on lib/engine/game/base.rb:
```diff
@@ -414,6 +414,7 @@ def new_stock_round
 
       def new_operating_round(round_num = 1)
         @log << "-- Operating Round #{@turn}.#{round_num} --"
+        @corporations.reject(&:floated?).each { |c| c.push_revenue_history!(nil, round_num) }
```
hm i'm not sure i like this

what if revenue history was more detailed

jcoveney replied:
Detailed in what way? I'm not sure what you mean. Do you mean if revenue history had turn and round information? Or...?

tobymao replied:
can we get around having to inject this fake data? what if the revenue history had more info so this didn't need to happen

jcoveney replied:
It could be a map keyed off of turn and round_num

tobymao replied:
yes, maybe let's try that

jcoveney replied:
ok, I'll get to that tomorrow. let me know if there are any big issues :) namely formatting of the cards lol

jcoveney commented:
Responded to a number of comments and added information to the cards. Note: it's not terribly attractive (I imagine you're sensing a theme). Where do you think it should go?

<img width="1033" alt="Screen Shot 2020-05-02 at 00 20 51" src="https://user-images.githubusercontent.com/830087/80822698-db4de480-8c0d-11ea-9739-44968a059c4b.png">

tobymao commented:
> Responded to a number of comments and added information to the cards. Note: it's not terribly attractive (I imagine you're sensing a theme). Where do you think it should go?
> 
> <img alt="Screen Shot 2020-05-02 at 00 20 51" width="1033" src="https://user-images.githubusercontent.com/830087/80822698-db4de480-8c0d-11ea-9739-44968a059c4b.png">

looks really bad. does it fit in the Header? or next to the president? I would label it Last Run: $...

jcoveney commented:
I was imagining having the information for the last OR (so, 1-3 pieces). If it's just 1 (eg the very last run), then yeah it can be "Last Run" and we can maybe put it next to the president, or right under it. I don't think there is space on the header.

jcoveney commented:
Current outstanding issues

- Refactor so that the information is map based, so that we don't have to inject nils
- figure out how to get the information on the cards

jcoveney commented:
Ok, I cleaned up the revenue history, got rid of the nils etc.

I went ahead and removed the information from the card. I really think that can/should be a different issue -- the main issue is not technical (it's not trivial to get the information), it's more of a design issue.

tobymao commented:
i'm going to hold off on this a bit because i'm reworking routes and revenues

tobymao commented:
well on second thought maybe we should add it now because people want it.

i'd like for it to be on the card, can you just put it to the left of the president, and say like 'Last Revenue'

tobymao commented on assets/app/view/spreadsheet.rb:
```diff
@@ -31,15 +31,44 @@ def render_table
+      end
+      or_hist_list = []
+      or_hist.keys.each { |x| or_hist_list << x }
+      or_hist_list.sort
```
how does this work with unfloated companies?

tobymao commented on assets/app/view/spreadsheet.rb:
```diff
@@ -31,15 +31,44 @@ def render_table
+      or_hist = {}
+      corporations.each do |c|
+        c.revenue_history.keys.each { |x| or_hist[x] = true }
+      end
```
ok i see what this is doing now, a bit convoluted, you can just do

corporations.flat_map { |c| c.revenue_history.keys }.uniq.sort

jcoveney replied:
Ah Ruby. "uniq". Of course. My Google Fu failed me earlier -- thanks!

jcoveney commented:
I should be able to get it updated tomorrow (China time)

I'll try putting it on the card next to president.

jcoveney commented:
Ok, was able to get it added to a card in what looks to be a non-hideous way. Take a look!

<img width="1036" alt="Screen Shot 2020-05-03 at 16 46 32" src="https://user-images.githubusercontent.com/830087/80909809-c41e0c80-8d5d-11ea-8c15-c06af06457bc.png">
